### PR TITLE
Add target-penalty annealing support (`penalty_schedule`) to the training loops

### LINKED ISSUE #41: Target-penalty annealing: a constant slack-penalty weight either hides target violations or degrades the learned policy
## Summary

Training with a **constant** target-slack penalty weight puts the user in a lose-lose
situation: a low/moderate constant weight produces policies whose targets are not
actually followable stage-by-stage (hidden by deterministic-equivalent evaluation),
while a high constant weight from epoch 1 destroys learning. **Annealing the penalty
multiplier during training fixes both.** Independent of model/example details, we
believe the training loop (or at least the examples) should support and default to an
annealed penalty schedule.

## Setup

- `examples/HydroPowerModels`, Bolivia ACP case, 96 stages, LSTM policy (2×128).
- Penalty expressed as `multiplier × auto_base`, where `auto_base` = largest absolute
  objective coefficient of the formulation (≈ 6000 for this case); applied to both the
  L1 and L2 slack terms of the target constraints.
- Each checkpoint evaluated two ways on 40 fixed scenarios (see companion issue):
  **Eval-A** = full-horizon deterministic equivalent with the policy's targets;
  **Eval-B** = stage-wise subproblem rollout (deployment semantics).
- Metric: objective **excluding** the target-slack penalty term ("operational cost"),
  plus a target-violation share.

## Symptom (single runs; det-eq training unless noted)

| penalty schedule (× auto_base) | Eval-A cost (violation) | Eval-B cost (violation) | B − A |
|---|---|---|---|
| constant ×1 (`:auto`) | 306,556 (0.046) | **324,437 (0.408)** | **+5.8%** |
| constant ×0.1 | 302,865 (0.005) | 320,449 (0.080) | +5.8% |
| constant ×30 | 370,464 (0.093) | 370,463 (0.093) | ≈0, but cost **+20.7%** |
| **anneal ×0.1→1→10→30** | **306,912.35 (0.0008)** | **306,912.33 (0.0008)** | **0.02** |
| constant ×1, stage-wise training | 356,186 (0.025) | 366,541 (0.010) | +2.9%, cost ~+19% |
| **anneal, stage-wise training** | **310,940.8 (0.0035)** | **310,941.5 (0.0035)** | **0.69** |

With a constant moderate weight, the policy looks excellent under the deterministic-
equivalent view but is +5.8% worse and violates its own targets ~41% of the time when
executed stage by stage. With a constant ×30 the two views agree but the policy is
~21% worse than achievable. Annealing ×0.1→×1→×10→×30 (we used phase lengths
proportional to 2/2/4/16 over 24 epochs; a compressed 2/2/2/2 over 8 epochs behaves
the same) yields policies that are simultaneously consistent (A ≡ B to ~1e-7 relative)
and at the good cost level, for **both** det-eq and stage-wise training.

## Why it happens

The full-horizon coupled solve can re-time and absorb unfollowable targets via slack,
so under a weak penalty the training signal never forces stage-wise-followable
targets; under a strong penalty from scratch, followability dominates the loss before
the policy has learned anything about cost.

## Proposed fix

Add a penalty-schedule option to `train_multistage` (and `train_multiple_shooting`) —
e.g., a vector of `(epoch_range, multiplier)` applied to an auto-calibrated base —
defaulting to the annealed ladder above, with constant weights still available
explicitly.

### PR DESCRIPTION
**Pull request by DJLee1208**
Fixes #41

## Summary

Training with a constant target-slack penalty weight either hides target violations (low/moderate weight:
the deterministic-equivalent view looks great while the stage-wise rollout violates its own targets) or
degrades the learned policy (high weight from epoch 1). This PR adds the penalty-schedule option proposed
in #41: a multiplier ladder ×0.1 → ×1 → ×10 → ×30 applied during training, relative to the penalty the
models were built with (so with `penalty=:auto`, relative to the auto-calibrated base
= max |objective coefficient|).

## What changed

- `src/utils.jl`: `default_annealed_schedule(num_batches)` (phase lengths proportional to 2/2/4/16, every
  phase ≥ 1 batch, < 4 batches falls back to the tail of the ladder so runs always end at ×30), schedule
  validation (contiguous from batch 1, finite positive multipliers, past-end hold), and internal helpers
  that capture the objective coefficients of the `norm_deficit` variables once at training start and
  rescale them at phase boundaries. Scaling that single coefficient covers all three `create_deficit!`
  modes (L1-only, L2-only, and the combined mode where the L1/L2 weights live in the linking constraint),
  as well as the renamed copies inside deterministic-equivalent and shooting-window models — no model
  rebuild is needed.
- `src/simulate_multistage.jl` / `src/multiple_shooting.jl`: a new `penalty_schedule` keyword on both
  `train_multistage` overloads and on `train_multiple_shooting`. `nothing` (default) leaves behavior
  exactly as today; `:default_annealed` builds the ladder over `num_batches`; an explicit vector of
  `(first_batch, last_batch, multiplier)` phases is validated and used as given.
- `examples/HydroPowerModels`: the three trainers default to the annealed schedule. They now make a single
  training call spanning `num_epochs*num_batches` batches so the schedule covers the whole run. Deliberate
  example-behavior notes: this keeps one optimizer state for the whole run (previously `Flux.setup` re-ran
  every epoch, silently resetting Adam moments), makes a `true` return from the record callback stop the
  whole run (previously it only ended the current epoch), and drops the `iter % 2100` batch-doubling hook
  (which the collapsed iteration count would have newly activated). Constant-penalty training remains
  available via `penalty_schedule = nothing`.
- `test/runtests.jl`: schedule-resolution exactness (including the 24-batch reference ladder
  `1:2 ×0.1, 3:4 ×1, 5:8 ×10, 9:24 ×30`), validation rejections, past-end hold, coefficient scaling in all
  three `create_deficit!` modes, untouched coefficients on the default path, and an end-to-end
  `train_multistage` run crossing a phase boundary on DiffOpt models.

## Validation

- `Pkg.test()` green on this branch (210/210, includes the new testsets).
- Reduced-scale demo (Bolivia DCP, 12 stages, 240 batches per arm, vanilla Ipopt, identical seeds and a
  fixed 6-scenario held-out set): with the constant as-built penalty the trained policy ends at a
  stage-wise-rollout target-violation share of **0.220** — far outside the ≤ ~0.05 trust gate — while its
  deterministic-equivalent view looks unremarkable; with the annealed default at the same budget the share
  drops to **0.052** — near but still slightly above the ≤ ~0.05 gate (4.2× lower). This is partial,
  reduced-scale evidence for the core mechanism of the issue (constant penalties learn targets that are
  not followable stage by stage; annealing restores followability); it is not reduced-scale gap closure.
  The full gap-closure endpoint of the issue table (det-eq ≡ rollout to ~1e-7 relative) requires
  full-scale training and is documented in the issue itself.

## Notes

- Library behavior is backward compatible: `penalty_schedule=nothing` is the default and the gated code
  path leaves every objective coefficient untouched. Only the examples change their default to annealed.
- The combined L1+L2 mode anneals both weights by the same multiplier (the ladder is a uniform multiplier
  by construction); annealing them separately is out of scope.
- Only `default_annealed_schedule` is exported; the capture/apply helpers are internal.
- Note: this PR and #41/#42 both touch
  `examples/HydroPowerModels/train_dr_hydropowermodels_subproblems.jl`; whichever merges second will be
  rebased onto the other.

### CODE REVIEW

File: src/simulate_multistage.jl [hunk 1]
```diff
@@ -339,19 +339,31 @@ function sample(uncertainty_samples::Vector{Vector{Tuple{VariableRef, Vector{T}}
     [sample(uncertainty_samples[t]) for t in 1:length(uncertainty_samples)]
 end
 
-function train_multistage(model, initial_state, subproblems::Vector{JuMP.Model}, 
-    state_params_in, state_params_out, uncertainty_sampler; 
+function train_multistage(model, initial_state, subproblems::Vector{JuMP.Model},
```
Comment (andrewrosemberg) [inline src/simulate_multistage.jl:342]: This is on me, but we should start adding nice docstrings to explain functions
